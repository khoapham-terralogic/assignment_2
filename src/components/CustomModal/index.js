import React, { useState, useEffect } from 'react';
import { Modal, ModalBody, ModalHeader, Button } from 'reactstrap';
import { toast } from 'react-toastify';

import MyDropzone from '../MyDropZone';

const CustomModal = ({ isOpen, toggle, uploadImage, modalCallback }) => {
    const [file, setFile] = useState(null)
    const callback = data => {
        setFile(data[0])
    }
    const handleSubmit = async (e) => {
        e.preventDefault();
        await uploadImage({ file })
        setFile(null)
        modalCallback(false)
    }
    useEffect(() => {
        if (file) {
            const fileType = file['type'];
            const validImageTypes = ['image/gif', 'image/jpeg', 'image/png'];
            if (!validImageTypes.includes(fileType)) {
                // invalid file type code goes here.
                toast.error("Invalid file type")
                setFile(null)
            }
        }
    }, [file])
    return (
        <Modal centered className="custom-modal" isOpen={isOpen} toggle={toggle}>
            <ModalHeader className="custom-modal__header" toggle={toggle}>
                <div className="custom-modal__header__title">
                    Upload image
                </div>
            </ModalHeader>
            <ModalBody className="custom-modal__body">
                <div className="custom-modal__body__container">
                    <MyDropzone parentCallback={callback} isActive={file} />
                    {/* {file ? <>
                        {isLoading ? <Button onClick={handleSubmit} disabled className="btn"><ClipSpinner color="#fff" size={20} /></Button> : <Button onClick={handleSubmit} className="btn">Upload</Button>}
                    </> : null} */}
                    {file ? <Button onClick={handleSubmit} className="btn">Upload</Button> : null}

                </div>
            </ModalBody>
        </Modal>
    );
}

export default CustomModal;
