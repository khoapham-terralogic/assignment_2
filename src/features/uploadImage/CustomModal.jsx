import React, { useState, useEffect } from 'react';
import { Modal, ModalBody, ModalHeader, Button } from 'reactstrap';
import { Form } from 'reactstrap'
import MyDropzone from './MyDropZone';
import store from '../../redux/store';
import { uploadImage } from '../../redux/actions/userAction'

const CustomModal = ({ isOpen, toggle }) => {
    const [file, setFile] = useState(null)
    const [isActive, setIsActive] = useState(false)
    const callback = data => {
        setFile(data[0])
    }
    const handleSubmit = (e) => {
        e.preventDefault();
        store.dispatch(uploadImage({ file }))
    }
    return (
        <Modal className="custom-modal" isOpen={isOpen} toggle={toggle}>
            <ModalHeader className="custom-modal__header" toggle={toggle}>
                <div className="custom-modal__header__title">
                    Upload image
                </div>
            </ModalHeader>
            <ModalBody className="custom-modal__body">
                <Form onSubmit={handleSubmit}>
                    <MyDropzone parentCallback={callback} isActive={file} />
                    <Button className="btn">Upload</Button>
                </Form>
            </ModalBody>
        </Modal>
    );
}

export default CustomModal;
