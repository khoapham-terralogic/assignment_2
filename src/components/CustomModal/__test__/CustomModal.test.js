import React from 'react'
import CustomModal from '..'
import { shallow, mount } from 'enzyme';
// import MyDropZone from '../MyDropZone'

describe('Test <CustomModal />', () => {
    let props = {
        isOpen: false,
        toggle: jest.fn(),
        uploadImage: jest.fn(),
        modalCallback: jest.fn()
    }
    // const file = new File([''], 'filename.txt', {
    //     type: 'text/plain',
    //     lastModified: new Date()
    // })
    const wrapper = shallow(<CustomModal {...props} />)
    it('should render', () => {
        expect(wrapper).toMatchSnapshot()
    });
    it('set file should be called when click upload', () => {
        // const changeState = jest.fn()
        // const wrapper = shallow(<CustomModal onClick={changeState} {...props} />)
        // const handleClick = jest.spyOn(React, 'useState')
        // handleClick.mockImplementation(file => [file, changeState]);
        // wrapper.find('.btn').simulate('click')
        // expect(changeState).toBeTruthy()

    })
    it('should not render button if file is empty', () => {
        // const file = new File([''], 'filename.txt', {
        //     type: 'text/plain',
        //     lastModified: new Date()
        // })
        let useEffect
        const mockUseEffect = () => {
            useEffect.mockImplementationOnce(f => f());
        };
        beforeEach(() => {
            useEffect = jest.spyOn(React, "useEffect").mockImplementation(f => f());
            mockUseEffect();
            mockUseEffect();
            const wrapper = shallow(<CustomModal {...props} />);
            expect(wrapper.find(".btn")).toHaveLength(0)
        });
    })
    // it('should call submit if button is click', () => {
    //     const wrapper = mount(<CustomModal {...props} />)
    //     wrapper.instance().setFile("somefile")
    //     console.log(wrapper.debug())
    //     // expect(handleSubmit).toHaveBeenCalled()
    // })
    it('should render button if file is not empty', () => {

        let useEffect
        const mockUseEffect = () => {
            useEffect.mockImplementationOnce(f => f());
        };
        beforeEach(() => {
            useEffect = jest.spyOn(React, "useEffect").mockImplementation(f => f());
            const stateSetter = jest.fn()
            jest.spyOn(React, 'useState')
                //Simulate that mode state value was set to 'new mode value'
                .mockImplementation(stateValue => [stateValue = {}, stateSetter])
            mockUseEffect();
            mockUseEffect();
            const wrapper = shallow(<CustomModal {...props} />);
            expect(wrapper.find(".btn")).toHaveLength(1)
        });
        // expect(wrapper.props()).toBe({})
    })
})
