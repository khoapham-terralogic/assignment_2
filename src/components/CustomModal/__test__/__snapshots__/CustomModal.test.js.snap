// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <CustomModal /> should render 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransition={
    Object {
      "mountOnEnter": true,
      "timeout": 150,
    }
  }
  centered={true}
  className="custom-modal"
  container="body"
  fade={true}
  isOpen={false}
  keyboard={true}
  modalTransition={
    Object {
      "timeout": 300,
    }
  }
  onClosed={[Function]}
  onOpened={[Function]}
  returnFocusAfterClose={true}
  role="dialog"
  scrollable={false}
  toggle={[MockFunction]}
  unmountOnClose={true}
  zIndex={1050}
>
  <ModalHeader
    charCode={215}
    className="custom-modal__header"
    closeAriaLabel="Close"
    tag="h5"
    toggle={[MockFunction]}
    wrapTag="div"
  >
    <div
      className="custom-modal__header__title"
    >
      Upload image
    </div>
  </ModalHeader>
  <ModalBody
    className="custom-modal__body"
    tag="div"
  >
    <div
      className="custom-modal__body__container"
    >
      <MyDropzone
        isActive={null}
        parentCallback={[Function]}
      />
    </div>
  </ModalBody>
</Modal>
`;
